### PR TITLE
Adding a search API for users and packages

### PR DESCRIPTION
**Pull request by akariv**

### CODE REVIEW

File: conductor/blueprints/search/models.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        return None
+    try:
+        if type(size) is list:
+            size = size[0]
+        api_params = {'size': int(size)}
+        api_params.update(kind_params)
+        q_str = ") AND (".join([" OR ".join(["{0}:{1}".format(k, json.dumps(json.loads(v)))
+                                             for v in v_arr])
+                                for k, v_arr in kw.items()
+                                if k.split('.')[0] in kind_params['_source']])
```
Comment (vitorbaptista) [inline conductor/blueprints/search/models.py]: Although array comprehensions lead to a smaller code, I tend to use them only for simple cases. It's hard to follow what's happening here.

Also, make sure that both keys and values are quoted, so you don't end up searching for `name:foo bar` when you really wanted `name:"foo bar"` or even `name:(foo bar)` (i.e. `name:foo name:bar`).

File: conductor/blueprints/search/models.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    return _engine
+
+
+def query(kind, size=100, **kw):
+    kind_params = ENABLED_SEARCHES.get(kind)
+    if kind_params is None:
+        return None
+    try:
+        if type(size) is list:
+            size = size[0]
```
Comment (vitorbaptista) [inline conductor/blueprints/search/models.py]: When `size` can be a list? I usually don't comment much, but I think it would be useful here.

File: conductor/blueprints/search/models.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        api_params = {'size': int(size)}
+        api_params.update(kind_params)
+        q_str = ") AND (".join([" OR ".join(["{0}:{1}".format(k, json.dumps(json.loads(v)))
+                                             for v in v_arr])
+                                for k, v_arr in kw.items()
+                                if k.split('.')[0] in kind_params['_source']])
+        if len(q_str) == 0:
+            q_str = "{0}:*".format(kind_params['_source'][0])
```
Comment (vitorbaptista) [inline conductor/blueprints/search/models.py]: I'm not sure if it'll work in your case, but an empty string (or even `undefined`) returns everything when using the elasticsearch JS library. If it's the same for Python, then it'll allow you to remove (or simplify) this line.

File: conductor/blueprints/search/models.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        ret = _get_engine().search(**api_params)
+        if ret.get('hits') is not None:
+            return [ hit['_source'] for hit in ret['hits']['hits'] ]
+        return None
+    except NotFoundError:
+        return None
+    except json.JSONDecodeError:
+        return None
+    except ValueError:
+        return None
```
Comment (vitorbaptista) [inline conductor/blueprints/search/models.py]: This is more about programming styles than anything else, so feel free to ignore.

Given that you're simply ignoring all these exceptions, I would merge them all in a single `except` clause. I would also remove the `return None` on line 51. All in all, apart from validating that the received params are good (as you do on lines 32~33), I prefer to, when possible, keep a single return point in every function. I feel like it's easier to read that way, but again, it's a matter of personal preferences.

**Comment by vitorbaptista** [comment]
@akariv Apart from the failing tests and my comments inline, it looks good to go :+1:

File: conductor/blueprints/search/models.py (hunk 1 above)
Comment (akariv) [inline conductor/blueprints/search/models.py]: 1. Agreed
2. Thanks for spotting the bug, I'll add a test for that

File: conductor/blueprints/search/models.py (hunk 2 above)
Comment (akariv) [inline conductor/blueprints/search/models.py]: Agreed

File: conductor/blueprints/search/models.py (hunk 1 above)
Comment (akariv) [inline conductor/blueprints/search/models.py]: BTW, since we're serializing to json everything before putting in the query, a situation like you described can't really happen,

File: conductor/blueprints/search/models.py (hunk 4 above)
Comment (akariv) [inline conductor/blueprints/search/models.py]: Okay
